File: data_structures/3_python/n_disjointSets.py
class DisjointSets:
    def __init__(self):
        self.parent = {}
        self.rank = {}

    def make_set(self, element):
        if element not in self.parent:
            self.parent[element] = element
            self.rank[element] = 0

    def find(self, element):
        if self.parent[element] == element:
            return element

        self.parent[element] = self.find(self.parent[element])  # Path compression
        return self.parent[element]

    def union(self, element1, element2):
        parent1 = self.find(element1)
        parent2 = self.find(element2)

        if parent1 != parent2:
            rank1 = self.rank.get(parent1, 0)
            rank2 = self.rank.get(parent2, 0)

            if rank1 > rank2:
                self.parent[parent2] = parent1
            elif rank1 < rank2:
                self.parent[parent1] = parent2
            else:
                self.parent[parent2] = parent1
                self.rank[parent1] = rank1 + 1

    def contains(self, element):
        return element in self.parent

    def set_size(self, element):
        representative = self.find(element)
        return sum(1 for val in self.parent.values() if self.find(val) == representative)

    def elements_in_same_set(self, element):
        representative = self.find(element)
        return [key for key, val in self.parent.items() if self.find(val) == representative]

    def all_sets(self):
        sets = {}
        set_representatives = set()

        for element in self.parent:
            representative = self.find(element)
            if representative not in set_representatives:
                set_elements = [key for key, val in self.parent.items() if self.find(val) == representative]
                sets[representative] = set_elements
                set_representatives.add(representative)

        return list(sets.values())

    def reset(self):
        self.parent = {}
        self.rank = {}

    def is_same_set(self, element1, element2):
        parent1 = self.find(element1)
        parent2 = self.find(element2)
        return parent1 == parent2

    def remove_set(self, element):
        representative = self.find(element)
        self.parent = {key: val for key, val in self.parent.items() if self.find(val) != representative}
        self.rank = {key: val for key, val in self.rank.items() if self.find(val) != representative}

    def path_to_root(self, element):
        path = []
        current_node = element
        representative = self.find(element)

        while current_node != representative:
            path.append(current_node)
            parent = self.find(current_node)
            current_node = parent

        path.append(representative)
        return path

    def get_representatives(self):
        representatives_set = set()
        representatives = []

        for element in self.parent:
            representative = self.find(element)
            if representative not in representatives_set:
                representatives.append(representative)
                representatives_set.add(representative)

        return representatives

    def find_and_compress(self, element):
        representative = self.find(element)
        self.compress_path(element, representative)
        return representative

    def compress_path(self, element, representative):
        current_node = element

        while current_node != representative:
            parent = self.find(current_node)
            self.parent[current_node] = representative
            current_node = parent

    def print_sets(self):
        sets = {}

        for element in self.parent:
            representative = self.find(element)
            if representative not in sets:
                sets[representative] = []

            sets[representative].append(element)

        for representative, elements in sets.items():
            print(f"Set with representative {representative}: {elements}")


# Example usage
disjoint_sets = DisjointSets()

disjoint_sets.make_set(1)
disjoint_sets.make_set(2)
disjoint_sets.make_set(3)
disjoint_sets.make_set(4)

disjoint_sets.union(1, 2)
disjoint_sets.union(3, 4)
disjoint_sets.union(2, 3)

print("Is 1 in the same set as 2:", disjoint_sets.is_same_set(1, 2))  # True
print("Is 1 in the same set as 3:", disjoint_sets.is_same_set(1, 3))  # True
print("Is 1 in the same set as 4:", disjoint_sets.is_same_set(1, 4))  # True
print("Is 2 in the same set as 4:", disjoint_sets.is_same_set(2, 4))  # True
print("Is 1 in the same set as 5:", disjoint_sets.is_same_set(1, 5))  # False

disjoint_sets.remove_set(3)

print("Is 1 in the same set as 2:", disjoint_sets.is_same_set(1, 2))  # True
print("Is 1 in the same set as 3:", disjoint_sets.is_same_set(1, 3))  # False
print("Is 1 in the same set as 4:", disjoint_sets.is_same_set(1, 4))  # True
print("Is 2 in the same set as 4:", disjoint_sets.is_same_set(2, 4))  # True

disjoint_sets.print_sets()

File: data_structures/2_rust/n_disjointSets.rs
use std::collections::{HashMap, HashSet};

#[derive(Debug, Eq, PartialEq, Hash, Clone, Copy)]
struct DisjointSets<T: Eq + PartialEq + Hash + Clone + Copy> {
    parent: HashMap<T, T>,
    rank: HashMap<T, i32>,
}

impl<T: Eq + PartialEq + Hash + Clone + Copy> DisjointSets<T> {
    fn new() -> Self {
        DisjointSets {
            parent: HashMap::new(),
            rank: HashMap::new(),
        }
    }

    fn make_set(&mut self, element: T) {
        if !self.parent.contains_key(&element) {
            self.parent.insert(element, element);
            self.rank.insert(element, 0);
        }
    }

    fn find(&mut self, element: T) -> Option<T> {
        let parent = *self.parent.get(&element).unwrap_or(&element);

        if parent == element {
            Some(parent)
        } else {
            let representative = self.find(parent)?;
            self.parent.insert(element, representative);
            Some(representative)
        }
    }

    fn union(&mut self, element1: T, element2: T) {
        let parent1 = self.find(element1).unwrap_or(element1);
        let parent2 = self.find(element2).unwrap_or(element2);

        if parent1 != parent2 {
            let rank1 = *self.rank.get(&parent1).unwrap_or(&0);
            let rank2 = *self.rank.get(&parent2).unwrap_or(&0);

            if rank1 > rank2 {
                self.parent.insert(parent2, parent1);
            } else if rank1 < rank2 {
                self.parent.insert(parent1, parent2);
            } else {
                self.parent.insert(parent2, parent1);
                self.rank.insert(parent1, rank1 + 1);
            }
        }
    }

    fn contains(&self, element: T) -> bool {
        self.parent.contains_key(&element)
    }

    fn set_size(&self, element: T) -> usize {
        let representative = self.find(element).unwrap_or(element);
        self.parent.values().filter(|&&val| self.find(val) == Some(representative)).count()
    }

    fn elements_in_same_set(&self, element: T) -> Vec<T> {
        let representative = self.find(element).unwrap_or(element);
        self.parent.keys().filter(|&&key| self.find(key) == Some(representative)).cloned().collect()
    }

    fn all_sets(&self) -> Vec<Vec<T>> {
        let mut sets: Vec<Vec<T>> = Vec::new();
        let mut set_representatives: HashSet<T> = HashSet::new();

        for &element in self.parent.keys() {
            let representative = self.find(element).unwrap();
            if !set_representatives.contains(&representative) {
                let set: Vec<T> = self.parent.keys()
                    .filter(|&&key| self.find(key) == Some(representative))
                    .cloned().collect();
                sets.push(set);
                set_representatives.insert(representative);
            }
        }

        sets
    }

    fn reset(&mut self) {
        self.parent.clear();
        self.rank.clear();
    }

    fn is_same_set(&mut self, element1: T, element2: T) -> bool {
        let parent1 = self.find(element1).unwrap_or(element1);
        let parent2 = self.find(element2).unwrap_or(element2);
        parent1 == parent2
    }

    fn remove_set(&mut self, element: T) {
        let representative = self.find(element).unwrap();
        self.parent.retain(|_, &mut val| self.find(val) != Some(representative));
        self.rank.retain(|_, &mut val| self.find(val) != Some(representative));
    }

    fn path_to_root(&mut self, element: T) -> Vec<T> {
        let mut path = Vec::new();
        let mut current_node = element;
        let representative = self.find(element).unwrap();

        while current_node != representative {
            path.push(current_node);
            let parent = self.find(current_node).unwrap();
            current_node = parent;
        }
        path.push(representative);
        path
    }

    fn get_representatives(&mut self) -> Vec<T> {
        let mut representatives_set: HashSet<T> = HashSet::new();
        let mut representatives: Vec<T> = Vec::new();

        for &element in self.parent.keys() {
            let representative = self.find(element).unwrap();
            if !representatives_set.contains(&representative) {
                representatives.push(representative);
                representatives_set.insert(representative);
            }
        }

        representatives
    }

    fn find_and_compress(&mut self, element: T) -> Option<T> {
        let representative = self.find(element)?;
        self.compress_path(element, representative);
        Some(representative)
    }

    fn compress_path(&mut self, element: T, representative: T) {
        let mut current_node = element;
        while current_node != representative {
            let parent = self.find(current_node).unwrap();
            self.parent.insert(current_node, representative);
            current_node = parent;
        }
    }

    fn print_sets(&self) {
        let mut sets: HashMap<T, Vec<T>> = HashMap::new();

        for &element in self.parent.keys() {
            let representative = self.find(element).unwrap();
            sets.entry(representative)
                .or_insert_with(Vec::new)
                .push(element);
        }

        for (representative, elements) in sets {
            println!("Set with representative {:?}: {:?}", representative, elements);
        }
    }
}

fn main() {
    // Example usage
    let mut disjoint_sets = DisjointSets::new();

    disjoint_sets.make_set(1);
    disjoint_sets.make_set(2);
    disjoint_sets.make_set(3);
    disjoint_sets.make_set(4);

    disjoint_sets.union(1, 2);
    disjoint_sets.union(3, 4);
    disjoint_sets.union(2, 3);

    println!("Is 1 in the same set as 2: {}", disjoint_sets.is_same_set(1, 2)); // true
    println!("Is 1 in the same set as 3: {}", disjoint_sets.is_same_set(1, 3)); // true
    println!("Is 1 in the same set as 4: {}", disjoint_sets.is_same_set(1, 4)); // true
    println!("Is 2 in the same set as 4: {}", disjoint_sets.is_same_set(2, 4)); // true
    println!("Is 1 in the same set as 5: {}", disjoint_sets.is_same_set(1, 5)); // false

    disjoint_sets.remove_set(3);

    println!("Is 1 in the same set as 2: {}", disjoint_sets.is_same_set(1, 2)); // true
    println!("Is 1 in the same set as 3: {}", disjoint_sets.is_same_set(1, 3)); // false
    println!("Is 1 in the same set as 4: {}", disjoint_sets.is_same_set(1, 4)); // true
    println!("Is 2 in the same set as 4: {}", disjoint_sets.is_same_set(2, 4)); // true

    disjoint_sets.print_sets();
}
